backend/updater.js

```diff
@@ -0,0 +1,247 @@
+        }).getHash();
+
+        // Remove this one from the hash map
+        sectionHashMap[sectionHash] = false;
+      }
+    }
+
+
+    // Find the sections that are still around
+    for (const sectionHash of Object.keys(sectionHashMap)) {
+      // If it was set to false, ignore it
+      if (!sectionHashMap[sectionHash]) {
+        continue;
+      }
+
+
+      macros.error('Section', sectionHash, "is being watched but it's class is not being watched?", sectionHashMap);
+    }
+
+    let allParsersOutput = [];
+
+    // Scrape the latest data
+    for (const aClass of classes) {
+      const latestData = await ellucianCatalogParser.main(aClass.prettyUrl);
+
+      allParsersOutput = allParsersOutput.concat(latestData);
+    }
+
+    const rootNode = {
+      type: 'ignore',
+      deps: allParsersOutput,
+      value: {},
+    };
+
+
+    // Because ellucianCatalogParser returns a list of classes, instead of a singular class, we need to run it on all of them
+    const output = await classesScrapers.runProcessors(rootNode);
+
+
+    // Keep track of which messages to send which users.
+    // The key is the facebookMessengerId and the value is a list of messages.
+    const userToMessageMap = {};
+
+    for (const aNewClass of output.classes) {
+      const hash = Keys.create(aNewClass).getHash();
+
+      const oldClass = this.dataLib.getClassServerDataFromHash(hash);
+
+      // Count how many sections are present in the new but not in the old.
+      let count = 0;
+      for (const crn of aNewClass.crns) {
+        if (!oldClass.crns.includes(crn)) {
+          count++;
+        }
+      }
+
+      let message = '';
+      const classCode = `${aNewClass.subject} ${aNewClass.classId}`;
+
+      if (count === 1) {
+        message = `A section was added to ${classCode}!`;
+      } else if (count > 1) {
+        message = `${count} sections were added to ${classCode}!`;
+      }
+
+      if (message) {
+        message += ` Check it out at https://searchneu.com/${aNewClass.subject}${aNewClass.classId} !`;
+
+        const user = classHashToUsers[hash];
+
+        if (!userToMessageMap[user]) {
+          userToMessageMap[user] = [];
+        }
+
+        userToMessageMap[user].push(message);
+      }
+    }
+
+
+    for (const newSection of output.sections) {
+      const hash = Keys.create(newSection).getHash();
+
+      const oldSection = this.dataLib.getSectionServerDataFromHash(hash);
+
+      // This should never run.
+      // The user should not be able to sign up for a section that didn't exist when they were signing up.
+      if (!oldSection) {
+        macros.error('Section was addded?', hash);
+        continue;
+      }
+
+      let message;
+
+      if (newSection.seatsRemaining > 0 && oldSection.seatsRemaining <= 0) {
+        message = `A seat opened up in ${newSection.subject} ${newSection.classId} (CRN: ${newSection.crn}). Check it out at https://searchneu.com/${newSection.subject}${newSection.classId} !`;
```

Comment:
should be
```
...newSection.classId}!`
```
not
```
...newSection.classId} !`
```

Reply:
I'd rather remove the space too, but on mobile the Facebook Messenger app will actually include the exclamation mark in the hyperlink. If we add a space only the `https://searchneu.com/CS2500` part will be hyperlinked.

Reply:
Oddly enough, on desktop, the exclamation mark is not included in the hyperlink.